File: frontend/src/lib/health/stats.ts
export const MIN_SAMPLE_SIZE = 10;
export const MIN_STD_THRESHOLD = 1e-6;
export const MAD_SCALE_FACTOR = 1.4826;

export function sumOrNull(values: number[]): number | null {
  return values.length > 0 ? values.reduce((a, b) => a + b, 0) : null;
}

export function meanOrNull(values: number[]): number | null {
  if (values.length === 0) return null;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

export function calculatePercentile(
  sortedValues: number[],
  percentile: number,
): number {
  if (sortedValues.length === 0) return 0;
  const index = (percentile / 100) * (sortedValues.length - 1);
  const lower = Math.floor(index);
  const upper = Math.ceil(index);
  if (lower === upper) return sortedValues[lower];
  return (
    sortedValues[lower] +
    (sortedValues[upper] - sortedValues[lower]) * (index - lower)
  );
}

export function winsorize(
  values: number[],
  lowerPercentile: number = 5,
  upperPercentile: number = 95,
): number[] {
  if (values.length < 4) return values;
  const sorted = [...values].sort((a, b) => a - b);
  const lowerBound = calculatePercentile(sorted, lowerPercentile);
  const upperBound = calculatePercentile(sorted, upperPercentile);
  return values.map((v) => Math.min(Math.max(v, lowerBound), upperBound));
}

export function calculateMedian(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : (sorted[mid - 1] + sorted[mid]) / 2;
}

export function calculateMAD(values: number[]): number {
  if (values.length < 2) return 0;
  const median = calculateMedian(values);
  const absoluteDeviations = values.map((v) => Math.abs(v - median));
  return calculateMedian(absoluteDeviations);
}

export interface RobustStats {
  median: number;
  mad: number;
  scaledMAD: number;
  mean: number;
  std: number;
}

export function calculateRobustStats(values: number[]): RobustStats {
  if (values.length === 0) {
    return { median: 0, mad: 0, scaledMAD: 0, mean: 0, std: 0 };
  }
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance =
    values.length > 1
      ? values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) /
        (values.length - 1)
      : 0;
  const std = Math.sqrt(variance);
  const median = calculateMedian(values);
  const mad = calculateMAD(values);
  return {
    median,
    mad,
    scaledMAD: MAD_SCALE_FACTOR * mad,
    mean,
    std,
  };
}

export function calculateStd(values: number[]): number {
  if (values.length < 2) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance =
    values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) /
    (values.length - 1);
  return Math.sqrt(variance);
}

export function calculateEMAValue(
  values: number[],
  span: number,
): number | null {
  if (values.length === 0) return null;

  const k = 2 / (span + 1);
  let ema = values[0];

  for (let i = 1; i < values.length; i++) {
    ema = values[i] * k + ema * (1 - k);
  }

  return ema;
}

File: src/analytics/stats.py
from __future__ import annotations

import math

import numpy as np
from scipy import stats as scipy_stats
from scipy.stats import median_abs_deviation
from scipy.stats.mstats import winsorize as _scipy_winsorize

from .constants import MAD_SCALE_FACTOR


def mean_or_none(values: list[float]) -> float | None:
    if not values:
        return None
    return float(np.mean(values))


def sum_or_none(values: list[float]) -> float | None:
    return float(np.sum(values)) if values else None


def calculate_percentile(sorted_values: list[float], percentile: float) -> float:
    if not sorted_values:
        return 0.0
    return float(np.percentile(sorted_values, percentile, method="linear"))


def winsorize(
    values: list[float], lower_pct: float = 5, upper_pct: float = 95
) -> list[float]:
    if len(values) < 4:
        return list(values)
    result = _scipy_winsorize(
        np.array(values), limits=(lower_pct / 100.0, 1.0 - upper_pct / 100.0)
    )
    return list(result.astype(float))


def calculate_median(values: list[float]) -> float:
    if not values:
        return 0.0
    return float(np.median(values))


def calculate_mad(values: list[float]) -> float:
    if len(values) < 2:
        return 0.0
    return float(median_abs_deviation(values, scale=1.0))


def calculate_std(values: list[float]) -> float:
    if len(values) < 2:
        return 0.0
    return float(np.std(values, ddof=1))


def calculate_robust_stats(values: list[float]) -> dict:
    if not values:
        return {"median": 0.0, "mad": 0.0, "scaled_mad": 0.0, "mean": 0.0, "std": 0.0}
    arr = np.array(values, dtype=float)
    mean = float(np.mean(arr))
    std = float(np.std(arr, ddof=1)) if len(values) >= 2 else 0.0
    median = float(np.median(arr))
    mad = float(median_abs_deviation(arr, scale=1.0))
    return {
        "median": median,
        "mad": mad,
        "scaled_mad": MAD_SCALE_FACTOR * mad,
        "mean": mean,
        "std": std,
    }


def calculate_ema_value(values: list[float], span: int) -> float | None:
    if not values:
        return None
    alpha = 2.0 / (span + 1)
    ema = values[0]
    for v in values[1:]:
        ema = alpha * v + (1 - alpha) * ema
    return float(ema)


def pearson_correlation(x: list[float], y: list[float]) -> float | None:
    r, _ = pearson_correlation_with_pvalue(x, y)
    return r


def pearson_correlation_with_pvalue(
    x: list[float],
    y: list[float],
) -> tuple[float | None, float | None]:
    from .constants import MIN_CORRELATION_PAIRS, MIN_STD_THRESHOLD

    n = len(x)
    if n != len(y) or n < MIN_CORRELATION_PAIRS:
        return None, None
    if np.std(x) < MIN_STD_THRESHOLD or np.std(y) < MIN_STD_THRESHOLD:
        return None, None
    try:
        result = scipy_stats.pearsonr(x, y)
        r_f = float(result[0])  # type: ignore[arg-type]  # scipy stubs
        p_f = float(result[1])  # type: ignore[arg-type]  # scipy stubs
        r_val = r_f if math.isfinite(r_f) else None
        p_val = p_f if math.isfinite(p_f) else None
        return r_val, p_val
    except Exception:
        return None, None
